drivers/usbdev/cdcncm.c

```diff
@@ -0,0 +1,2541 @@
+ * Name: cdcncm_setconfig
+ *
+ *   Set the device configuration by allocating and configuring endpoints and
+ *   by allocating and queue read and write requests.
+ *
+ ****************************************************************************/
+
+static int cdcncm_setconfig(FAR struct cdcncm_driver_s *self, uint8_t config)
+{
+  struct usb_epdesc_s epdesc;
+  bool is_high_speed = self->usbdev.speed == USB_SPEED_HIGH;
+  int ret = OK;
+
+  if (config == self->config)
+    {
+      return OK;
+    }
+
+  cdcncm_resetconfig(self);
+
+  if (config == CDCECM_CONFIGID_NONE)
+    {
+      return OK;
+    }
+
+  if (config != CDCECM_CONFIGID)
+    {
+      return -EINVAL;
+    }
+
+  cdcncm_mkepdesc(CDCNCM_EP_INTIN_IDX, &epdesc, &self->devinfo, false);
+  ret = EP_CONFIGURE(self->epint, &epdesc, false);
+
+  if (ret < 0)
+    {
+      goto error;
+    }
+
+  self->epint->priv = self;
+  cdcncm_mkepdesc(CDCNCM_EP_BULKIN_IDX,
+                  &epdesc, &self->devinfo, is_high_speed);
+  ret = EP_CONFIGURE(self->epbulkin, &epdesc, false);
+
+  if (ret < 0)
+    {
+      goto error;
+    }
+
+  self->epbulkin->priv = self;
+
+  cdcncm_mkepdesc(CDCNCM_EP_BULKOUT_IDX,
+                  &epdesc, &self->devinfo, is_high_speed);
+  ret = EP_CONFIGURE(self->epbulkout, &epdesc, true);
+
+  if (ret < 0)
+    {
+      goto error;
+    }
+
+  self->epbulkout->priv = self;
+
+  /* Queue read requests in the bulk OUT endpoint */
+
+  DEBUGASSERT(!self->rxpending);
+
+  self->rdreq->callback = cdcncm_rdcomplete,
+  ret = EP_SUBMIT(self->epbulkout, self->rdreq);
+  if (ret != OK)
+    {
+      uerr("EP_SUBMIT failed. ret %d\n", ret);
+      goto error;
+    }
+
+  /* We are successfully configured */
+
+  self->config = config;
+
+  /* Set client's MAC address */
+
+  memcpy(self->dev.netdev.d_mac.ether.ether_addr_octet,
+         "\x00\xe0\xde\xad\xbe\xef", IFHWADDRLEN);
+
+  /* Report link up to networking layer */
+
+  cdcncm_ifup(&self->dev);
+
+  return OK;
+
+error:
+  cdcncm_resetconfig(self);
+  return ret;
+}
+
+/****************************************************************************
+ * Name: ncm_notify
+ *
+ ****************************************************************************/
+
+static int ncm_notify(FAR struct cdcncm_driver_s *self)
+{
+  FAR struct usb_ctrlreq_s *req =
+                            (FAR struct usb_ctrlreq_s *)self->notifyreq->buf;
+  int ret = 0;
+
+  switch (self->notify)
+    {
+      case NCM_NOTIFY_NONE:
+        return ret;
+
+      case NCM_NOTIFY_CONNECT:
+
+        /* Notifying the host of the NIC modification status */
+
+        req->req      = NCM_NETWORK_CONNECTION;
+        req->value[0] = LSBYTE(IFF_IS_RUNNING(self->dev.netdev.d_flags));
+        req->value[1] = MSBYTE(IFF_IS_RUNNING(self->dev.netdev.d_flags));
+        req->len[0]   = 0;
+        req->len[1]   = 0;
+        ret           = sizeof(*req);
+
+        self->notify  = NCM_NOTIFY_NONE;
+        break;
+
+      case NCM_NOTIFY_SPEED:
+        {
```

Comment:
```suggestion
```

Reply:
`FAR uint32_t *data;`
1415 have variable definitions, do I need to delete it?

Reply:
ok, let us keep the bracket